### PR TITLE
lr scheduler implementation (on device training)

### PR DESCRIPTION
**Pull request by pengwa**
**Description**: lr scheduler implementations.

> Implemented linear lr scheduler
> Unit tests with test data generator. 
> Move test_runner.cc into trainer/trainer.cc. 

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: orttraining/orttraining/training_api/include/optimizer.h [hunk 1]
```diff
@@ -88,43 +90,9 @@ struct Optimizer {
...
-    ORT_NOT_IMPLEMENTED("Not implemented.");
-    return Status::OK();
-  }
-
- private:
-  float start_factor_;
-  float end_factor_;
-  int64_t total_iters_;
+  // Currently all parameters are in a single group, so we hardcode group0 here.
+  const std::string group_zero_name_ = "group0";
```
Comment (baijumeswani) [inline orttraining/orttraining/training_api/include/optimizer.h:95]: can this string literal be moved to an unnamed namespace in `optimizer.cc` file?

File: orttraining/orttraining/training_api/include/lr_scheduler.h [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+   * Must be implemented by sub classes.
+   */
+  virtual float ComputeLRMultiplicativeFactorInternal(int64_t step) = 0;
+};
+
+/**
+ * @brief Decays learning rate by linearly updated multiplicative factor.
+ *
+ */
+struct LinearLRScheduler : public MultiplicativeLRSchedulerBase {
```
Comment (ashbhandare) [inline orttraining/orttraining/training_api/include/lr_scheduler.h:74]: what other LR schedulers can inherit this intermediate class?

File: orttraining/orttraining/training_api/include/optimizer.h [hunk 3]
```diff
@@ -64,10 +64,12 @@ struct Optimizer {
...
   Status GetStateDict(OptimizerCheckpointState& optimizer_checkpoint_states);
 
+  Status LoadStateDict(OptimizerCheckpointState& optimizer_checkpoint_states);
+
   int64_t GetStep() const {
     return optimizer_state_.step;
   }
-  
+
   Status SetLearningRate(float lr) {
```
Comment (ashbhandare) [inline orttraining/orttraining/training_api/include/optimizer.h:73]: are we allowing users to call optimizer->SetLearningRate()? If not, need to make this method private

Comment (baijumeswani) [reply orttraining/orttraining/training_api/include/optimizer.h:73]: `GetStep` can also be private.

File: onnxruntime/test/testdata/test_data_generation/lr_scheduler/warmup_linear_scheduler_warmupstep-70_restored.json [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+            6.666666666666667e-05
+        ]
+    ],
+    [
+        99,
+        [
+            3.3333333333333335e-05
+        ]
+    ]
+]
```
Comment (baijumeswani) [inline onnxruntime/test/testdata/test_data_generation/lr_scheduler/warmup_linear_scheduler_warmupstep-70_restored.json:296]: newline.

File: orttraining/orttraining/training_api/include/lr_scheduler.h (hunk 2 above)
Comment (pengwa) [reply orttraining/orttraining/training_api/include/lr_scheduler.h:74]: There is a common kind of schedulers calculating lr this way:
lr = multiplicative_factor * initial_lr;
Different schedulers in the same kind, they define different functors for getting multiplicative_factor. 

So this class is for this usage as a base class. 


For other kind of schdulers, they need inherit from LRSchedulerBase

File: orttraining/orttraining/training_api/include/optimizer.h (hunk 3 above)
Comment (pengwa) [reply orttraining/orttraining/training_api/include/optimizer.h:73]: Fixed both.

File: orttraining/orttraining/training_api/include/lr_scheduler.h [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    return optim_.optimizer_state_.step;
+  }
+
+  float GetInitialLRInternal() {
+    return optim_.optimizer_state_.initial_lr;
+  }
+
+ private:
+  virtual float ComputeLearningRateInternal() = 0;
+  Optimizer& optim_;
```
[the remark this answers is not in the document]
Comment (pengwa) [reply orttraining/orttraining/training_api/include/lr_scheduler.h]: Fixed the first comment. For the second one, I think we would suggest user to create pointers if they want to do that way, while we need improve anyway on how we expose LRSchduler (and even Base) later to users, because ideally we should allow user to define their own lr schduler.

File: orttraining/orttraining/training_api/include/optimizer.h (hunk 1 above)
Comment (pengwa) [reply orttraining/orttraining/training_api/include/optimizer.h]: fixed.

File: onnxruntime/test/testdata/test_data_generation/lr_scheduler/warmup_linear_scheduler_warmupstep-70_restored.json (hunk 4 above)
Comment (pengwa) [reply onnxruntime/test/testdata/test_data_generation/lr_scheduler/warmup_linear_scheduler_warmupstep-70_restored.json:296]: this is auto generated by json.dump. I think it is fine. :)

File: orttraining/orttraining/training_api/include/lr_scheduler.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (baijumeswani) [reply orttraining/orttraining/training_api/include/lr_scheduler.h]: My first and second comment were kind of implying the same thing. If we require a shared pointer in LRScehduler, then the compiler can create a copy constructor for us for scheduler and so doing what I showed above would work.
Thanks for making the change.

**Comment by askhade** [comment]
@pengwa  : Can you remove the test_runner changes from this PR. I am updating the test_runner to use C API. Apart from this the rest of the changes look good to me.

**Comment by askhade** [comment]
> @pengwa : Can you remove the test_runner changes from this PR. I am updating the test_runner to use C API. Apart from this the rest of the changes look good to me.

On second thought - I think it is OK to keep your changes... I will make updates on top of your changes.

**Review by askhade — APPROVED** [review]
